### PR TITLE
Infrastructure class

### LINKED ISSUE #117: How to handle Vert.x instances - add Infrastructure class
We now have multiple service discovery providers using a Vert.x instance. At the moment each provider creates its own instance. There are a few problems with this approach:

* the instances are not closed - of course, Stork is intended to be closed when the JVM terminates, but in multi-app environments, it could be a problem;
* each instance has a cost, and lead to the creation of event loop threads

We would need a way for the providers to get access to a potentially managed instance of Vert.x to avoid the two issues.

### PR DESCRIPTION
**Pull request by michalszynkiewicz**
fixes #117

### CODE REVIEW

**Comment by michalszynkiewicz** [comment]
how quarkus integration looks like with this: https://github.com/michalszynkiewicz/quarkus/commit/820006edae65a0987b434cf5f4ee8d02afed348f

**Comment by cescoffier** [comment]
I like the approach.
The danger is that beans you will look up in Quarkus may have been removed. So, extensions will be required to be sure that the beans are not removed.

File: api/src/main/java/io/smallrye/stork/integration/DefaultStorkInfrastructure.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package io.smallrye.stork.integration;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+public class DefaultStorkInfrastructure implements StorkInfrastructure {
```
Comment (cescoffier) [inline api/src/main/java/io/smallrye/stork/integration/DefaultStorkInfrastructure.java:7]: Javadoc missing

File: api/src/main/java/io/smallrye/stork/integration/DefaultStorkInfrastructure.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    private final Map<Class<?>, Object> utilities = new ConcurrentHashMap<>();
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> T get(Class<T> utilityClass, Supplier<T> defaultSupplier) {
+        return (T) utilities.computeIfAbsent(utilityClass, key -> defaultSupplier.get());
```
Comment (cescoffier) [inline api/src/main/java/io/smallrye/stork/integration/DefaultStorkInfrastructure.java:13]: Probability check that utilityClass and defaultSupplier are not null.

File: api/src/main/java/io/smallrye/stork/integration/StorkInfrastructure.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package io.smallrye.stork.integration;
+
+import java.util.function.Supplier;
+
+public interface StorkInfrastructure {
+    <T> T get(Class<T> utilityClass, Supplier<T> defaultSupplier);
```
Comment (cescoffier) [inline api/src/main/java/io/smallrye/stork/integration/StorkInfrastructure.java:6]: Javadoc missing

It should specify what should happen if the default supplier throws a runtime exception

**Review by cescoffier — CHANGES_REQUESTED** [review]

**Review by aureamunoz — APPROVED** [review]
LGTM

File: api/src/main/java/io/smallrye/stork/integration/DefaultStorkInfrastructure.java (hunk 1 above)
Comment (michalszynkiewicz) [reply api/src/main/java/io/smallrye/stork/integration/DefaultStorkInfrastructure.java:12]: added

File: api/src/main/java/io/smallrye/stork/integration/DefaultStorkInfrastructure.java (hunk 2 above)
Comment (michalszynkiewicz) [reply api/src/main/java/io/smallrye/stork/integration/DefaultStorkInfrastructure.java:20]: done

File: api/src/main/java/io/smallrye/stork/integration/StorkInfrastructure.java (hunk 3 above)
Comment (michalszynkiewicz) [reply api/src/main/java/io/smallrye/stork/integration/StorkInfrastructure.java:26]: I added the javadoc. What happens with exceptions thrown by supplier depends on an implementation...

**Review by cescoffier — APPROVED** [review]

**Pull request merged**